sdk-rs/src/types.rs

```diff
@@ -0,0 +1,322 @@
+use std::cmp::Ordering;
+
+// re-export types in public API
+pub use drift_program::{
+    controller::position::PositionDirection,
+    state::{
+        order_params::{OrderParams, PostOnlyParam},
+        user::{MarketType, Order, OrderType, PerpPosition, SpotPosition},
+    },
+};
+use solana_sdk::{instruction::AccountMeta, pubkey::Pubkey};
+use thiserror::Error;
+
+use crate::constants::{perp_markets, spot_markets};
+
+/// Drift program context
+#[derive(Debug, Copy, Clone)]
+#[repr(u8)]
+pub enum Context {
+    /// Target Devenet
+    Dev,
+    /// Target Mainnet
+    Mainnet,
+}
+
+/// Id of a Drift market
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+pub struct MarketId {
+    pub(crate) index: u16,
+    pub(crate) kind: MarketType,
+}
+
+impl MarketId {
+    /// Lookup a market id by context and symbol
+    ///
+    /// This operation is not free so lookups should be reused/cached by the caller
+    ///
+    /// Returns an error if symbol and context do not map to a known market
+    pub fn lookup(context: Context, symbol: &str) -> Result<Self, ()> {
+        let mut parts = symbol.split('-');
+        match (parts.next(), parts.next()) {
+            (Some(base), None) => {
+                let markets = spot_markets(context);
+                if let Some(market) = markets.iter().find(|m| m.symbol.eq_ignore_ascii_case(base)) {
+                    return Ok(MarketId::spot(market.market_index));
+                }
+            }
+            (Some(base), Some(perp)) => {
+                if perp.eq_ignore_ascii_case("perp") {
+                    let markets = perp_markets(context);
+                    if let Some(market) = markets
+                        .iter()
+                        .find(|m| m.base_asset_symbol.eq_ignore_ascii_case(base))
+                    {
+                        return Ok(MarketId::perp(market.market_index));
+                    }
+                }
+            }
+            _ => (),
+        }
+
+        Err(())
+    }
+    /// Id of a perp market
+    pub const fn perp(index: u16) -> Self {
+        Self {
+            index,
+            kind: MarketType::Perp,
+        }
+    }
+    /// Id of a spot market
+    pub const fn spot(index: u16) -> Self {
+        Self {
+            index,
+            kind: MarketType::Spot,
+        }
+    }
+}
+
+impl From<(u16, MarketType)> for MarketId {
+    fn from(value: (u16, MarketType)) -> Self {
+        Self {
+            index: value.0,
+            kind: value.1,
+        }
+    }
+}
+
+/// Provides builder API for Orders
+#[derive(Default)]
+pub struct NewOrder {
```

Comment:
feel like this will just end up needing to expose all the same params as the OrderParams over time, why'd you decide to make it a separate struct?

Reply:
a distinct type for the builder is common but using OrderParams sounds good too, happy to make this change